### PR TITLE
[FLINK-34221][table-runtime] Introduce MiniBatchStreamingJoinOperator

### PR DESCRIPTION
**Pull request by xishuaidelin**
## What is the purpose of the change

This pull request introduce a new operator to support minibatch join. This operator use BufferBundle to store the records. 

## Brief change log

Introduce a operator called MiniBatchSteamingJoinOperator which implement the main logic of minibatch join. And 
 the records are copied before being placed into BufferBundle. Besides, the metric class SimpleGauge is introduced to collect the reduced size of records in the BufferBundle. 

## Verifying this change

Added  StreamingMinibatchOperatorTest to verify the change.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java [hunk 1]
```diff
@@ -0,0 +1,1667 @@
...
+                        "LineOrd#2",
+                        "4 Bellevue Drive, Pottstown, PB 19464",
+                        "Ord#5",
+                        "LineOrd#2",
+                        "TRUCK"));
+    }
+
+    @Test
+    @Tag("miniBatchSize=2")
+    @DisplayName("LeftJoinHasUniqueKeyRetAndAcc")
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java:1181]: We should add more tests to cover the -U & +U suppression logic, include the else code block logic.

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java [hunk 2]
```diff
@@ -0,0 +1,1667 @@
...
+                        "8 Bellevue Drive, Pottstown, PF 19464",
+                        "Ord#1",
+                        "LineOrd#1",
+                        "AIR"));
+    }
+
+    /** Outer join only emits INSERT or DELETE Msg. */
+    @Tag("miniBatchSize=10")
+    @Test
+    public void testLeftJoinWithJoinKeyContainsUniqueKey() throws Exception {
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java:904]: I think there should be some more tests for LEFT JOIN, including logic for record folding and suppression, mainly for some boundary cases.

**Review by lsyldliu — COMMENTED** [review]
@xishuaidelin Thanks for your contribution, I left some comments. Please fix the checkstyle

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java:1181]: @lsyldliu Yes, I hava added more tests to cover the else code block logic in suppression logic.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+            }
+        } else if (inputBuffer instanceof JoinKeyContainsUniqueKeyBundle) {
+            for (Map.Entry<RowData, List<RowData>> entry : inputBuffer.getRecords().entrySet()) {
+                // set state key first
+                setCurrentKey(entry.getKey());
+                Iterator<RowData> iter = entry.getValue().iterator();
+                processElementWithSuppress(
+                        iter, inputSideStateView, otherSideStateView, inputIsLeft);
+            }
+        } else {
```
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java]: Done.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/AbstractStreamingJoinOperator.java [hunk 4]
```diff
@@ -66,6 +67,10 @@ public abstract class AbstractStreamingJoinOperator extends AbstractStreamOperat
     protected transient JoinConditionWithNullFilters joinCondition;
     protected transient TimestampedCollector<RowData> collector;
 
+    protected transient boolean requiresCopy;
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/AbstractStreamingJoinOperator.java:70]: We don't need this member variable, this only used for minibatch join, it must need to copy.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+/** Streaming unbounded Join base operator which support mini-batch join. */
+public abstract class MiniBatchStreamingJoinOperator extends StreamingJoinOperator
+        implements BundleTriggerCallback {
+
+    private static final long serialVersionUID = -1106342589994963997L;
+
+    private final CoBundleTrigger<RowData, RowData> coBundleTrigger;
+
+    private transient BufferBundle<?> leftBuffer;
+    private transient BufferBundle<?> rightBuffer;
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:52]: protected transient TypeSerializer<RowData> leftSerializer;
    protected transient TypeSerializer<RowData> rightSerializer;

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+
+    @Override
+    public void open() throws Exception {
+        super.open();
+
+        // must copy in mini-batch
+        this.requiresCopy = true;
+
+        this.leftBuffer = initialBuffer(leftInputSideSpec);
+        this.rightBuffer = initialBuffer(rightInputSideSpec);
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:80]: this.leftSerializer = leftType.createSerializer(new ExecutionConfig());
        this.rightSerializer = rightType.createSerializer(new ExecutionConfig());

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+                .getMetricGroup()
+                .gauge("leftBundleReducedSize", leftBundleReducedSizeGauge);
+        getRuntimeContext()
+                .getMetricGroup()
+                .gauge("rightBundleReducedSize", rightBundleReducedSizeGauge);
+    }
+
+    @Override
+    public void processElement1(StreamRecord<RowData> element) throws Exception {
+        RowData record = leftSerializer.toSerializedRow(element.getValue(), requiresCopy);
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:100]:
```suggestion
        RowData record = leftSerializer.copy(element.getValue());
```

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+        if (leftInputSideSpec.getUniqueKeySelector() != null) {
+            uniqueKey = leftInputSideSpec.getUniqueKeySelector().getKey(record);
+        }
+        leftBuffer.addRecord(joinKey, uniqueKey, record);
+        coBundleTrigger.onElement1(record);
+    }
+
+    @Override
+    public void processElement2(StreamRecord<RowData> element) throws Exception {
+        RowData record = rightSerializer.toSerializedRow(element.getValue(), requiresCopy);
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:112]: ditto

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/AbstractRowDataSerializer.java [hunk 9]
```diff
@@ -33,6 +33,12 @@ public abstract class AbstractRowDataSerializer<T extends RowData> extends Paged
     /** Get the number of fields. */
     public abstract int getArity();
 
+    /** Convert a {@link RowData} to a serialized row. */
+    public RowData toSerializedRow(T rowData, boolean requiresCopy) throws IOException {
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/AbstractRowDataSerializer.java:37]: Revert this change, we don't need it.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+        super.processWatermark2(mark);
+    }
+
+    @Override
+    public void prepareSnapshotPreBarrier(long checkpointId) throws Exception {
+        finishBundle();
+    }
+
+    @Override
+    public void finish() throws Exception {
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:140]: ping again

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+        } else if (inputBuffer instanceof JoinKeyContainsUniqueKeyBundle) {
+            for (Map.Entry<RowData, List<RowData>> entry : inputBuffer.getRecords().entrySet()) {
+                // set state key first
+                setCurrentKey(entry.getKey());
+                Iterator<RowData> iter = entry.getValue().iterator();
+                processElementWithSuppress(
+                        iter, inputSideStateView, otherSideStateView, inputIsLeft);
+            }
+        } else {
+            for (RowData jk : inputBuffer.getJoinKeys()) {
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:242]: ping again

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperatorTestBase.java [hunk 12]
```diff
@@ -41,7 +41,7 @@
 /** Base test class for {@link AbstractStreamingJoinOperator}. */
 public abstract class StreamingJoinOperatorTestBase {
 
-    protected final InternalTypeInfo<RowData> leftTypeInfo =
+    protected InternalTypeInfo<RowData> leftTypeInfo =
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperatorTestBase.java:44]: The final keyword shouldn't be remove for leftTypeInfo

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperatorTestBase.java [hunk 13]
```diff
@@ -51,17 +51,17 @@ public abstract class StreamingJoinOperatorTestBase {
...
 
-    protected final InternalTypeInfo<RowData> rightTypeInfo =
+    protected InternalTypeInfo<RowData> rightTypeInfo =
             InternalTypeInfo.of(
                     RowType.of(
                             new LogicalType[] {new CharType(false, 20), new CharType(true, 10)},
                             new String[] {"line_order_id0", "line_order_ship_mode"}));
 
-    protected final RowDataKeySelector leftKeySelector =
+    protected RowDataKeySelector leftKeySelector =
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperatorTestBase.java:60]: ditto

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java [hunk 14]
```diff
@@ -0,0 +1,416 @@
...
+            extends MiniBatchStreamingJoinOperator {
+
+        public MiniBatchRightOuterJoinStreamOperator(MiniBatchStreamingJoinParameter parameter) {
+            super(parameter);
+        }
+
+        @Override
+        protected void processBundles(BufferBundle leftBuffer, BufferBundle rightBuffer)
+                throws Exception {
+
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:388]: ping again

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java [hunk 15]
```diff
@@ -0,0 +1,1807 @@
...
+                        "Ord#X",
+                        "LineOrd#2",
+                        "AIR"));
+    }
+
+    @Tag("miniBatchSize=1")
+    @Test
+    public void testFullJoinWithRightArriveFirstWithNoMiniBatch() throws Exception {
+        // joinKey is LineOrd
+        // left fold  || right fold
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java:269]: I think the above six tests doesn't test the fold logic, only for reducing retract message of timing-related left and right stream.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/AbstractStreamingJoinOperator.java (hunk 4 above)
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/AbstractStreamingJoinOperator.java]: Yes, this could be removed.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java (hunk 6 above)
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:80]: Yes, I have changed this and simplified the code.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java (hunk 7 above)
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java]: ditto

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperatorTestBase.java (hunk 12 above)
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperatorTestBase.java:44]: Yes, I changed this and only removed the final for rightType, rightKeySelector and assertor.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java (hunk 5 above)
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:53]: done

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java (hunk 8 above)
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java]: done

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:143]: Yes. I add this to clear the bundle.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java [hunk 16]
```diff
@@ -0,0 +1,416 @@
...
+            Iterator<RowData> iter,
+            JoinRecordStateView inputSideStateView,
+            JoinRecordStateView otherSideStateView,
+            boolean inputIsLeft)
+            throws Exception {
+        RowData pre = null; // always retractMsg if not null
+        while (iter.hasNext()) {
+            RowData current = iter.next();
+            boolean isSuppress = false;
+            if (RowDataUtil.isRetractMsg(current) && iter.hasNext()) {
```
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:185]: Yes, I added the logic in the bundle to keep last retract record when there are mulpile retract records in the bundle which has uniqueKey. So there would not be multiple retract records here.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:245]: Done.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java [hunk 17]
```diff
@@ -0,0 +1,416 @@
...
+            extends MiniBatchStreamingJoinOperator {
+
+        public MiniBatchLeftOuterJoinStreamOperator(MiniBatchStreamingJoinParameter parameter) {
+            super(parameter);
+        }
+
+        @Override
+        protected void processBundles(BufferBundle leftBuffer, BufferBundle rightBuffer)
+                throws Exception {
+            // more efficient to process right first for left out join, i.e, some retractions can be
```
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:369]: Done.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/MiniBatchStreamingJoinOperator.java:391]: Done

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperatorTestBase.java (hunk 13 above)
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingJoinOperatorTestBase.java:60]: Done

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java [hunk 18]
```diff
@@ -0,0 +1,1667 @@
...
+                        "3 Bellevue Drive, Pottstown, PA 19464",
+                        "Ord#X",
+                        "LineOrd#2",
+                        "AIR"));
+
+    }
+
+    @Tag("miniBatchSize=1")
+    @Test
+    public void testRightJoinWithRightArriveFirstWithNoMiniBatch() throws Exception {
```
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java:198]: Done

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java (hunk 15 above)
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java:269]: Yes, these tests are not supposed to test the fold logic. They are only designed to test the timing-related left and right streams which could avoid emiting redundant records.

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java (hunk 2 above)
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMiniBatchJoinOperatorTest.java:950]: Ok, I think update records with more rowkinds and disorder could be tested. I would improve the coverage of tests.

**Review by lsyldliu — APPROVED** [review]
LGTM
